### PR TITLE
8268978: Document the javadoc software stack

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
Please review a `noreg-doc` update to document the javadoc software stack, in the top-level `jdk.javadoc.internal` package.

The intent is to provide a broad overview.   We can add more specific details, but in general, additional details would probably be better in the `package-info` files for the constituent packages.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268978](https://bugs.openjdk.java.net/browse/JDK-8268978): Document the javadoc software stack

### Reviewers
 * [Hannes Wallnöfer](https://openjdk.java.net/census#hannesw) (@hns - **Reviewer**) ⚠️ Review applies to f33232518c6e95e1cd06b150ea433919c37ceda7

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6914/head:pull/6914` \
`$ git checkout pull/6914`

Update a local copy of the PR: \
`$ git checkout pull/6914` \
`$ git pull https://git.openjdk.java.net/jdk pull/6914/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6914`

View PR using the GUI difftool: \
`$ git pr show -t 6914`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6914.diff">https://git.openjdk.java.net/jdk/pull/6914.diff</a>

</details>

### CODE REVIEW

**Comment by hns** [comment]
A general question about this: Where and how will this documentation mostly be consumed? The `jdk.javadoc.internal` package is not included in standard JDK documentation, so is it likely that this documentation will be mainly consumed within an editor/IDE? 

My feeling is that maybe a less markup-heavy and more text centric form would be easier to consume in that context. For instance, using appropriate HTML headings and paragraphs instead of the deeply nested `<dl>` structure might provide more of a continuous text experience, while still maintaining the hierarchical information. Of course this is a matter of taste, different people have different preferences.

**Comment by jonathan-gibbons** [comment]
> A general question about this: Where and how will this documentation mostly be consumed? The `jdk.javadoc.internal` package is not included in standard JDK documentation, so is it likely that this documentation will be mainly consumed within an editor/IDE?
> 
> My feeling is that maybe a less markup-heavy and more text centric form would be easier to consume in that context. For instance, using appropriate HTML headings and paragraphs instead of the deeply nested `<dl>` structure might provide more of a continuous text experience, while still maintaining the hierarchical information. Of course this is a matter of taste, different people have different preferences.

This is to document the design for the benefit of potential future developers.  I have no illusions that we will ever run javadoc over the internal documentation as a matter of course, but it is getting better/easier to view documentation comments in an editor./IDE, and with that in mind, I note that IntelliJ IDEA does an excellent job in its Reader Mode of displaying the comment and handling the HTML markup.  I also think it is worth keeping relatively high standards for internal documentation, at least for internal public and protected API. (I accept that it is hard to persuade people to better comment private API :-( )

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/package-info.java [hunk 1]
```diff
@@ -24,21 +24,142 @@
...
+ *          the toolkit layer. In times past, it was common to write doclets to analyze
+ *          code using the then-current API as an early version of a Java language model.
+ *          That old API has been replaced by the {@link javax.lang.model Language Model API},
+ *          and tools that wish to use that API to analyze Java programs have a choice of
+ *          how to invoke it, using the <em>javac</em> support for
+ *          {@linkplain javax.annotation.processing annotation processing},
+ *          or {@linkplain com.sun.source.util.Plugin plugins}, as well as doclets.
+ *          Which is best for any application will depend of the circumstances, but
+ *          if a tool does not need access to the documentation comments in a program,
+ *          it is possible that one of the other invocation mechanisms will be more convenient.
```
Comment (hns) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/package-info.java:121]: Should the "Other Doclets" part above be added to the "Doclet Interface" section below? I think they deal with very related topics. The text above goes into some detail about when and how to implement doclets, which is the topic of the text below. Also, the list above mostly describes the actual components of the javadoc module, but "Other Doclets" is not really part of that.

**Review by hns — APPROVED** [review]
This is a very useful overview and introduction!

It looks very good to me, except I think the information in "Other Doclets" should be merged with or added to the "Doclet Interface" section as described below.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/package-info.java (hunk 1 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/package-info.java:121]: I'll take a look.

Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/package-info.java:121]: I think it helps to keep them separate ... and the text is adjacent.

"Other Doclets" is somewhat more about why you might not want to write a doclet these days ... it's addressing the historical past where there were lots of doclets, a lot of which did not even look at doc comments ... i.e they were using the Doclet API as an early Language Model API.

Comment (hns) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/package-info.java:121]: Fine with me. Your explanation makes sense, another bit of history learned.

Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/package-info.java:121]: I'd forgotten, until now, there is a more detailed version of this text on the OpenJDK website.  I'll leave all the text here, but add an `@see`

**Pull request closed without merging**